plugins/gopacket/monitor/pod.go

```diff
@@ -0,0 +1,166 @@
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/Portshift/go-utils/k8s"
+	log "github.com/sirupsen/logrus"
+	"github.com/up9inc/mizu/tap"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	ResyncPeriod       = 1 * time.Minute
+	Interval           = 5
+	KubernetesAPIQPS   = 80.0
+	KubernetesAPIBurst = 60
+)
+
+type PodMonitor struct {
+	clientset       *kubernetes.Clientset
+	stopCh          chan struct{}
+	podStore        cache.Store
+	isStateChange   bool
+	namespacesToTap map[string]bool
+	lock            sync.RWMutex
+}
+
+func NewPodMonitor(namespaces []string) (*PodMonitor, error) {
+	clientset, _, err := k8s.CreateK8sClientset(nil, k8s.KubeOptions{
+		KubernetesAPIQPS:   KubernetesAPIQPS,
+		KubernetesAPIBurst: KubernetesAPIBurst,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create K8s clientset: %v", err)
+	}
+	return &PodMonitor{
+		clientset:       clientset,
+		stopCh:          make(chan struct{}),
+		isStateChange:   true,
+		namespacesToTap: sliceToMap(namespaces),
+	}, nil
+}
+
+func sliceToMap(s []string) map[string]bool {
+	ret := make(map[string]bool)
+
+	for _, s2 := range s {
+		ret[s2] = true
+	}
+	return ret
+}
+
+func (pm *PodMonitor) getIsStateChange() bool {
+	pm.lock.RLock()
+	defer pm.lock.RUnlock()
+
+	return pm.isStateChange
+}
+
+func (pm *PodMonitor) setIsStateChange(val bool) {
+	pm.lock.Lock()
+	defer pm.lock.Unlock()
+
+	pm.isStateChange = val
+}
+
+func (pm *PodMonitor) Start(ctx context.Context) {
+	watchlist := cache.NewListWatchFromClient(pm.clientset.CoreV1().RESTClient(), v1.ResourcePods.String(), v1.NamespaceAll, fields.Everything())
+	store, controller := cache.NewInformer(
+		watchlist,
+		&v1.Pod{},
+		ResyncPeriod,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				pod := obj.(*v1.Pod) //nolint:forcetypeassert
+				if pm.namespacesToTap[pod.Namespace] {
+					pm.setIsStateChange(true)
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				pod := obj.(*v1.Pod) //nolint:forcetypeassert
+				if pm.namespacesToTap[pod.Namespace] {
+					pm.setIsStateChange(true)
+				}
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				oldPod := oldObj.(*v1.Pod) //nolint:forcetypeassert
+				newPod := newObj.(*v1.Pod) //nolint:forcetypeassert
+				if oldPod.Status.PodIP != newPod.Status.PodIP && pm.namespacesToTap[newPod.Namespace] {
+					pm.setIsStateChange(true)
+				}
+			},
+		},
+	)
+	pm.podStore = store
+
+	go controller.Run(pm.stopCh)
+
+	ticker := time.NewTicker(Interval * time.Second)
+	for {
+		select {
+		case <-ctx.Done():
+			log.Info("Done")
+			return
+		case <-ticker.C:
+			if pm.getIsStateChange() {
+				log.Info("Pod ips changed, updating filter authorities")
+				pm.updateFilteredAuthorities()
+			}
+		}
+	}
+}
+
+func (pm *PodMonitor) updateFilteredAuthorities() {
+	podIPs := pm.GetPodsIps()
+	tap.SetFilterAuthorities(podIPs)
+	pm.setIsStateChange(false)
```

Comment:
If I miss the state change I will catch up with it in the next interval (5 sec)

Reply:
you are setting it to false here `pm.setIsStateChange(false)` so you will miss it if no new changes will appear

Reply:
Are we sure that the store is in sync with the ResourceEventHandlerFuncs events? maybe an event was called but the srtore is not yet in sync? maybe better to get/set the value from the handlers?
Agree about the race that if in this function a handler sets it to true after line 136 this change theoretically will never/for a long time take place. Maybe it's better to listen on a change channel instead of doing it periodically?

Reply:
> Are we sure that the store is in sync with the ResourceEventHandlerFuncs events

From the docs "NewInformer returns a Store and a controller for populating the store while also providing event notifications.", so I think that we can be sure that if an event was called the store was populate.

> Maybe it's better to listen on a change channel instead of doing it periodically

I also think it would be better, with one extra thought about the first "scan". When the store is being populated the first time we will get lots of "Add" events, we should wait until the store is in synced (using `WaitForCacheSync`) and only then call `SetFilterAuthorities`, after that we can update filter authorities on any change.

Reply:
I don't think it is a good idea to do it on every event.
In the case of someone delete all pods, you will need to update the list of pods and setAuthorities a lot of times instead of waiting for it to finish

Reply:
I will add a lock in this function